File: ts/easy/mergeSort.ts
const seaArr = ['cod', 'herring', 'marling'];
const freshArr = ['asp', 'carp', 'ide', 'trout'];

function merge(sea: string[], fresh: string[]): string[] {
    const result: string[] = [];
    const copySea = [...sea];
    const copyFresh = [...fresh];
    const MOST_HEAVY_STR = 'zzzzzz';

    while (copySea.length || copyFresh.length) {
      const seaFish = copySea[0] || MOST_HEAVY_STR;
      const freshFish = copyFresh[0] || MOST_HEAVY_STR;

      if (seaFish < freshFish) {
        const fish = copySea.shift();
        if (fish) {
          result.push(fish);
        }
      } else if (seaFish > freshFish) {
        const fish = copyFresh.shift();
        if (fish) {
          result.push(fish);
        }
    }
    }
    return result;
}
console.log('Output: ', merge(seaArr, freshArr));

function mergeSort(arr: string[]): string[] {
  if (arr.length === 1) return arr;
  const mid = Math.floor(arr.length / 2);
  const left = arr.slice(0, mid);
  const right = arr.slice(mid);
  return merge(mergeSort(left), mergeSort(right))
}
console.log(mergeSort(['cod', 'herring', 'marling', 'asp', 'carp', 'ide', 'trout']));

File: python/easy/merge_sort.py
from typing import List
import math

sea_list = ["cod", "herring", "marling"]
fresh_list = ["asp", "carp", "ide", "trout"]

def merge(sea: List[str], fresh: List[str]) -> List[str]:
    result = []
    copy_sea = sea[::]
    copy_fresh = fresh[::]
    MOST_HEAVY_STR = 'zzzzzz'

    while len(copy_sea) > 0 or len(copy_fresh) > 0:
        sea_fish = MOST_HEAVY_STR
        fresh_fish = MOST_HEAVY_STR

        if len(copy_sea):
            sea_fish = copy_sea[0]
        if len(copy_fresh):
            fresh_fish = copy_fresh[0]

        if sea_fish < fresh_fish:
            fish = copy_sea.pop(0)
            if fish:
                result.append(fish)
        elif sea_fish > fresh_fish:
            fish = copy_fresh.pop(0)
            if fish:
                result.append(fish)
    return result
print(merge(sea_list, fresh_list))

def merge_sort(l):
    if len(l) == 1:
        return l
    mid = math.floor(len(l) / 2)
    left = l[0:mid]
    right = l[mid:]
    return merge(merge_sort(left), merge_sort(right))
print(merge_sort(["cod", "herring", "marling", "asp", "carp", "ide", "trout"]))
